RetailCoder.VBE/UI/DockableWindowHost.cs

```diff
@@ -7,86 +7,87 @@
+    // pointers (IntPtr) to pass around the COM interfaces, and so great care has to be taken with ensuring explicit reference counting is 
+    // COM compliant.
+    //
+    // CLASSES:
+    // _DockableWindowHost: the main class, implementing the proxy IOleObject and IOleInPlaceObject interfaces
+    // ExposedUserControl: exposes the internal IOleObject and IOleInPlaceObject interfaces from System.Windows.Forms.UnsafeNativeMethods
+    // the rest are wrapper classes and helpers.
+    // The old subclassing code is still used, at the bottom of the new _DockableWindowHost class.
+    //
+    // All defined IOle* interfaces are _compatible_ [not identical] to the originals.  Aggregation objects are used to pass interfaces 
+    // from our own versions of the IOle* interfaces to the UserControl ones (defined in System.Windows.Forms.UnsafeNativeMethods)
+    //  In particular, our interfaces differ by defining all pointers, including object instances, as IntPtr
+    //   this is for two reasons: 
+    //    1) we need to support nullified reference input params.  C# does not support them, and the CLR will throw exceptions with them.
+    //    2) we have to avoid the CLR resolving the aggregated objects to our own versions of the interfaces whenever possible
+    //   - this means that manual reference counting (with Marshal.AddRef / Release) is in place to ensure proper COM compliance
+    //
+    // Differences from the old DockableWindowHost:
+    //   we now have a proper IOleObject::Close() event that is the last thing to fire on our object.  We should use this to tear down
+    //   all the COM instances obtained for each instance of DockableWindowHost.  This is equivalent to the OnDisconnection event on the 
+    //   main AddIn side, and we should respect that the VBE expects us to release everything relating to the ActiveX control at this exact point.
+    //   IMO the main addin should NOT now trigger our destruction process.  It should be self contained, using the IOleObject::Close event.
+    //
+    // TODO improve comments
+    // TODO some C# magic?  (I only dabble in C#, so I'm sure this isn't going to win awards for the best use of C# features ;)
+    //
+    // -- Wayne Phillips 29th Dec 2017
+    //
+    //
+    // Note that the wrapper infrastructure has been moved to the namespace Rubberduck.VBEditor.CustomComWrappers.
+    //
+    //
+    
 
-        [StructLayout(LayoutKind.Explicit)]
-        private struct LParam
-        {
-            [FieldOffset(0)]
-            public uint Value;
-            [FieldOffset(0)]
-            public readonly ushort LowWord;
-            [FieldOffset(2)]
-            public readonly ushort HighWord;
-        }
+    // ExposedUserControl - wrapper for UserControl that also exposes the underlying 
+    // IOleObject and IOleInPlaceObject COM interfaces implemented by it
+    public class ExposedUserControl : UserControl
```

Comment:
I think this should be its own file.

Reply:
I would prefer to leave this in the same file, but at the bottom, until anything but the dockable host really needs an exposed user control. (I cannot really see what that would be.) This would follow the same reasoning as the `ParentWindow` class, which is also defined in this file.